packages/nextjs/src/config/loaders/ast.ts

```diff
@@ -0,0 +1,258 @@
+    //
+    // Note: If we were being super precise about this, we'd also check the context in which the identifier is being
+    // used, because there are some cases where we actually don't want to be renaming things (if the identifier is being
+    // used to name a class property, for example). But the chances that someone is going to have a class property in a
+    // nextjs page file with the same name as one of the canonical functions are slim to none, so for simplicity we can
+    // stop filtering here. If this ever becomes a problem, more precise filter checks can be found in a comment at the
+    // bottom of this file.
+    return Identifier.check(nodePath.node) && nodePath.node.name === name;
+  };
+
+  return identifierFilter;
+}
+
+/**
+ * Create a filter for nodes declaring variables with the given name
+ *
+ * @param name The variable name to filter on
+ * @returns A filter function with the name baked in
+ */
+function makeDeclarationFilter(name: string): ASTFilterFunction {
+  // Check for a structure of the form
+  //
+  //     node: VariableDeclaration
+  //      \
+  //       declarations: VariableDeclarator[]
+  //        \
+  //         0 : VariableDeclarator
+  //          \
+  //           id: Identifier
+  //            \
+  //             name: string
+  //
+  // where `name` matches the given name.
+  const declarationFilter = function (path: jscsTypes.ASTPath<ASTNodeType>): boolean {
+    return (
+      VariableDeclaration.check(path.node) &&
+      path.node.declarations.length === 1 &&
+      VariableDeclarator.check(path.node.declarations[0]) &&
+      Identifier.check(path.node.declarations[0].id) &&
+      path.node.declarations[0].id.name === name
+    );
+  };
+
+  return declarationFilter;
+}
+
+/**
+ * Create a filter for nodes representing exports with the given name
+ *
+ * @param name The variable name to filter on
+ * @returns A filter function with the name baked in
+ */
+function makeExportFilter(name: string): ASTFilterFunction {
+  const exportFilter = function (path: jscsTypes.ASTPath<ASTNodeType>): boolean {
+    return ExportSpecifier.check(path.node) && path.node.exported.name === name;
+  };
+
+  return exportFilter;
+}
+
+/**
+ * Find all nodes of the given type with the given name
+ *
+ * @param ast The code, in AST form
+ * @param nodeType The type of node for which to search
+ * @param name The identifier which should be associated with the found nodes
+ * @returns A collection of NodePaths pointing to any nodes which were found
+ */
+function findNodes(ast: jscsTypes.Collection, nodeType: string, name: string): jscsTypes.Collection {
+  const filterFactories: { [key: string]: (name: string) => ASTFilterFunction } = {
+    VariableDeclaration: makeDeclarationFilter,
+    Identifier: makeIdentifierFilter,
+    ExportSpecifier: makeExportFilter,
+  };
+  const filter = filterFactories[nodeType](name);
+  return ast.find(Node).filter(filter);
+}
```

Comment:
I don't understand exactly why we need the factory function pattern here? To me it seems like we could simply declare the filters above w/o the factory and then collect them in the object in this function?

Reply:
To me this seems a bit overcomplicated. Can't we just do the filtering in `findIdentifiers`, `findDeclarations` and `findExports` respectively.

Reply:
We can, and that's honestly originally how I had it. I changed it a) in order to break up what were becoming long functions (some of which have ended up simplified in the meantime, to be fair), and b) in order to be able to unify it into a single function (`findNodes`) since I was ending up doing the same thing in all of them.

Lemme see what in-between I can find, if any.

Reply:
Nice!